src/fasta.rs

```diff
@@ -1,209 +1,932 @@
+}
 
-#[derive(Error, Debug)]
-pub enum DnaFastaParseError {
-    #[error("could not translate FASTA contents")]
-    TranslationError(#[from] TranslationError),
-    #[error("could not read FASTA")]
-    IOError(#[from] std::io::Error),
+pub struct FastaParser<T: FromStr> {
+    settings: FastaParseSettings,
+    _marker: std::marker::PhantomData<T>,
 }
 
-impl DnaFastaParser {
-    pub fn parse<R: Read>(handle: R) -> Result<Vec<(String, DnaSequence)>, DnaFastaParseError> {
-        SimpleFastaParser::parse(handle)?
-            .into_iter()
-            .map(|(title, sequence)| -> Result<(String, DnaSequence), _> {
-                let dna = DnaSequence::from_str(&*sequence)?;
-                Ok((title, dna))
-            })
-            .collect()
+impl<T: FromStr> FastaParser<T> {
+    /// Construct a new FastaParser with the given [`FastaParseSettings`]
+    pub fn new(settings: FastaParseSettings) -> Self {
+        Self {
+            settings,
+            _marker: Default::default(),
+        }
+    }
+
+    /// Construct a new FastaParser with strict settings (see [`FastaParseSettings::strict()`])
+    pub fn strict() -> Self {
+        Self::new(FastaParseSettings::strict())
+    }
+
+    /// Construct a new FastaParser with lax settings (see [`FastaParseSettings::lax()`])
+    pub fn lax() -> Self {
+        Self::new(FastaParseSettings::lax())
+    }
+
+    pub fn parse<R: BufRead>(
+        &self,
+        handle: R,
+    ) -> Result<Vec<FastaRecord<T>>, FastaParseError<T::Err>> {
+        let mut result: Vec<FastaRecord<T>> = vec![];
+        let mut state = ParserState::StartOfFile {
+            contents: "".to_string(),
+        };
+
+        let mut line_number = 0;
+        for (idx, line) in handle.lines().enumerate() {
+            let line = line?;
+            line_number = idx + 1;
+
+            let (new_state, record) = state.advance_line(&self.settings, &line, line_number);
+            state = new_state;
+            if let Some(record) = record {
+                result.push(record.parse().map_err(FastaParseError::ParseError)?);
+            }
+        }
+
+        if let Some(record) = state.advance_eof(&self.settings, line_number + 1) {
+            result.push(record.parse().map_err(FastaParseError::ParseError)?);
+        }
+
+        Ok(result)
+    }
+
+    pub fn parse_str(&self, s: &str) -> Result<Vec<FastaRecord<T>>, FastaParseError<T::Err>> {
+        self.parse(s.as_bytes())
+    }
+}
+
+/// Try to parse a FASTA header (prefixed with > or ;), returning the line without the prefix char, and with any trailing
+/// newlines removed
+fn try_parse_header(line: &str) -> Option<&str> {
+    let head = line.chars().next();
+    // ; is semi-obsolete alternative header char
+    if head == Some('>') || head == Some(';') {
+        Some(line[1..].trim_end_matches('\n')) // we know it's an ASCII char so this slice is panic-safe
+    } else {
+        None
     }
 }
 
 #[cfg(test)]
 mod tests {
```

Comment:
oh man this is just screaming for `quickcheck` tests

Reply:
Hmm, maybe? I don't see how this could fail / what quickcheck would uncover (I guess that's the point of property testing :-) ). I actually realized in the latest commit that the `trim_end_matches` here is pointless, so the line is just `Some(&line[1..])` now, which I really really don't think can fail :-)

Reply:
Oh, I didn't mean this specific snippet, just this module in general.